pkg/archive: add xattr to archive when use overlay2 graphdriver

coolljt0725 opened the pull request:
When use overlay2 as the graphdriver and the kernel enable
`CONFIG_OVERLAY_FS_REDIRECT_DIR=y`, the following dockerfile
will miss files after mv dir operation.

```
FROM busybox

RUN mkdir /dir1
RUN touch /dir1/newfile
RUN mv /dir1 /dir2
```
The `newfile` is missing from the `/dir2` in the final image.

This is because `dir2` has a xattr to redirect its dir to `dir1`
when kernel enbale `CONFIG_OVERLAY_FS_REDIRECT_DIR`
```
debugfs:  ea_list dir2/
Extended attributes:
  security.selinux (43)
  trusted.overlay.redirect (4) = "dir1"
```
but this xattr is missing on archiving.
```
debugfs:  ea_list dir2
Extended attributes:
  security.selinux (43)
```

This patch add `Xattrs` and `XattrsInclude` to TarOptions and
enable this two options when using overlay2.

Signed-off-by: Lei Jitang <leijitang@huawei.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

coolljt0725 commented:
ping @dmcgowan  @stevvooe  @rhvgoyal

rhvgoyal commented:
Right, with `CONFIG_OVERLAY_FS_REDIRECT_DIR=y` overlay saves an xattr in upper directory inode which contains path to directory in lower/. And if that xattr is lost, contents of renamed directory will be lost. Before this feature, overlay will deny renaming of non-empty directories.

So conceptually it makes sense to save this xattr and restore.

rhvgoyal commented:
cc @amir73il 

Amir introduced this feature in kernel. CCing him.

AkihiroSuda commented:
This change seems image layers unportable to non-overlay hosts?

rhvgoyal commented:
Not just non-overlay hosts, what about hosts having overlay with CONFIG_OVERLAY_FS_REDIRECT_DIR=n``

thaJeztah commented:
also ping @ijc @dmcgowan @justincormack (IIRC, this came up recently for recent debian kernels)

coolljt0725 committed:
pkg/archive: add xattr to archive when use overlay2 graphdriver

When use overlay2 as the graphdriver and the kernel enable
`CONFIG_OVERLAY_FS_REDIRECT_DIR=y`, the following dockerfile
will miss files after mv dir operation.

```
FROM busybox

RUN mkdir /dir1
RUN touch /dir1/newfile
RUN mv /dir1 /dir2
```
The `newfile` is missing from the `/dir2` in the final image.

This is because `dir2` has a xattr to redirect its dir to `dir1`
when kernel enbale `CONFIG_OVERLAY_FS_REDIRECT_DIR`
```
debugfs:  ea_list dir2/
Extended attributes:
  security.selinux (43)
  trusted.overlay.redirect (4) = "dir1"
```
but this xattr is missing on archiving.
```
debugfs:  ea_list dir2
Extended attributes:
  security.selinux (43)
```

This patch add `Xattrs` and `XattrsInclude` to TarOptions and
enable this two options when using overlay2.

Signed-off-by: Lei Jitang <leijitang@huawei.com>

dmcgowan commented:
This is not the right solution. This feature needs to be detected and fallback to using naive diff. The other possible solution would require the archive code to interpret the xattr and perform the rename in outputted tar as we do for the whiteout flags. I am not supportive of the latter approach and see the current method of altering output based on overlay specific flags as error prone, the best example being this feature not working in overlay2 and any future kernel development on overlay.

AkihiroSuda commented:
Closing in favor of #34342

amir73il commented:
Just wanted to add that going forward, i.e. kernel V4.12, it's not advised to preserve trusted.overlay.origin xattr, so that's another good reason to nack this PR.
